File: categoryTheory/Chapter2/memoize.gpt.ts
type UnaryFunction<T, R> = (x: T) => R;

function memoize<T, R>(fn: UnaryFunction<T, R>): UnaryFunction<T, R> {
    const cache = new Map<T, R>();

    return function(x: T): R {
        if (!cache.has(x)) {
            cache.set(x, fn(x));
        }
        return cache.get(x)!;
    }
}

File: categoryTheory/Chapter2/memoize.gpt.hs
import qualified Data.Map as Map

memoize :: (Ord a) => (a -> b) -> a -> b
memoize f = let cache = Map.empty in \x ->
    case Map.lookup x cache of
        Just y  -> y
        Nothing -> let y = f x in y `seq` Map.insert x y cache `seq` y
